src/kern/npf_ctl.c

```diff
@@ -732,6 +758,70 @@ npfctl_save(npf_t *npf, u_long cmd, void *data)
 	return error;
 }
 
+/*
+ * npfctl_table_replace_nvlist: atomically replace a table's contents
+ * with the passed table data.
+ */
+static int __noinline
+npfctl_table_replace_nvlist(npf_t *npf, nvlist_t *npf_dict, nvlist_t *errdict)
+{
+	npf_tableset_t *tblset;
+	int error = 0;
+
+	npf_table_t *oldt, *newt;
+
+	npf_config_enter(npf);
+	tblset = npf_config_tableset(npf);
+
+	/* Get the entries or binary data. */
+	error = npf_mk_table(npf, npf_dict, errdict, tblset, &newt, 1);
+	if (error) {
+		goto err_config;
+	}
+
+	oldt = npf_tableset_swap(tblset, newt);
+	if (oldt == NULL) {
+		error = EINVAL;
+		goto err_config;
+	}
+	npf_config_sync(npf);
+	npf_config_exit(npf);
+
+	/* At this point, it is safe to destroy the old table. */
+	npf_table_destroy(oldt);
+
+	/* successfully completed */
+	goto out;
```

Comment:
This `goto` is redundant and actually makes things more convoluted than necessary. How about:
```c
	npf_table_t *tbl, *gc_tbl;

	...

	gc_tbl = npf_tableset_swap(tblset, tbl);
	if (gc_tbl == NULL) {
		error = EINVAL;
		gc_tbl = tbl;
		goto err;
	}
	npf_config_sync(npf);
err:
	npf_config_exit(npf);
	npf_table_destroy(gc_tbl);
	return error;
```
So you have the `gc_tbl` variable which is either old or new (depending whether there is an error) to garbage collect.

Reply:
Great suggestion, that's a much cleaner structure. I've changed this locally too.